src/service/invoice-pdf-service.ts

```diff
@@ -0,0 +1,236 @@
+/**
+ *  SudoSOS back-end API service.
+ *  Copyright (C) 2020  Study association GEWIS
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Affero General Public License as published
+ *  by the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+import Invoice from '../entity/invoices/invoice';
+import {
+  Address,
+  Client,
+  Company,
+  Dates,
+  FileResponse,
+  FileSettings,
+  Identity,
+  IInvoiceRouteParams,
+  InvoiceParameters,
+  InvoiceReferences,
+  InvoiceRouteParams,
+  InvoiceType,
+  Language,
+  Product,
+  ProductPricing,
+  ReturnFileType,
+  TotalPricing,
+  VAT,
+} from 'pdf-generator-client';
+import InvoicePdf from '../entity/file/invoice-pdf';
+import FileService from './file-service';
+import InvoiceEntry from '../entity/invoices/invoice-entry';
+import { hashJSON } from '../helpers/hash';
+
+export interface PdfGenerator {
+  client: Client,
+  fileService: FileService
+}
+
+// Used for grouping in the PDF.
+// These are 'hardcoded' since if these would change the template also would have to change.
+const PDF_VAT_ZERO = 0;
+const PDF_VAT_LOW = 9;
+const PDF_VAT_HIGH = 21;
+
+const UNUSED_PARAM = '';
+
+export default class InvoicePdfService {
+
+  /**
+   * Validates whether the PDF hash of an invoice matches the expected hash. This is used to determine if the invoice's PDF representation has been altered or needs regeneration.
+   *
+   * @param {Invoice} invoice - The invoice entity to validate the PDF hash for.
+   * @returns {boolean} Returns `true` if the PDF hash matches the expected hash, indicating no changes; returns `false` otherwise.
+   */
+  static validatePdfHash(invoice: Invoice): boolean {
+    if (!invoice.pdf) return false;
```

Comment:
If statement not covered (but should be an easy fix).

Reply:
fixed